### PR TITLE
fix TSInterfaceDeclaration, TSInterfaceBody, TSInterfaceHeritage

### PR DESCRIPTION
**Pull request by despairblue**

### CODE REVIEW

File: tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -44,8 +51,17 @@ interface Derived extends Base, Base2 {
...
 
-interface Derived2 extends Base, Base2 { // error
-  x: { a: number, b: string };
+// comment above TSInterfaceDeclaration
+interface Derived2 extends Base, Base2 // comment above TSInterfaceDeclaration
+{
+  // comment next TSInterfaceDeclaration
+  // comment above TSTypeLiteral
+  // comment inside of ObjectTypeAnnotation
+  // comment below TSTypeLiteral
```
Comment (despairblue) [inline tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap]: All comments get attached to the `TSInterfaceDeclaration` thus they all end up in one place. I think that used to work correctly.

@JamesHenry is this intentional? Then I would need to go through the whole body and use the location to attach the comments to the corresponding node before printing the body.

File: src/printer.js [hunk 2]
```diff
@@ -762,6 +762,14 @@ function genericPrintNoParens(path, options, print, args) {
       ]);
     }
     case "TSInterfaceDeclaration":
+      // Attach comments to the TSInterfaceBody since we want to print them
```
Comment (vjeux) [inline src/printer.js]: Can you do that inside of `comments.js`, we do this kind of things a lot where we attach comments to different nodes

File: tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (vjeux) [inline tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap]: Prettier implements its own logic to place comments, it is not eslint-typescript related. What may trigger this is the location of nodes being incorrect.

**Comment by vjeux** [comment]
Could you split this one into two, the comment logic which is not ready to land, and the other part which is?

**Comment by despairblue** [comment]
@vjeux done

File: tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (despairblue) [inline tests/typescript/conformance/interfaces/interfaceDeclarations/__snapshots__/jsfmt.spec.js.snap]: @vjeux 

Mhh, I find it weird that for iterfaces all comments are always attached to the `TSInterfaceDeclaration`, whereas for classes and objects the comments are attached to the following node:

```ts
interface I {
  a: string;
  // this comment is attached to the `TSInterfaceDeclaration` node which is right under `Program`
  b: string;
}
```

```ts
class C {
  public a: string;
  // this comment is attached to the following `ClassProperty` node
  public b: string;
}
```

```ts
const c = {
  a: 1,
  // this comment attached to the following `Property` node
  b: 2,
}
```

To me it seems as if the printer decides where to put the comments. And for that is uses mostly the logic to print the comment before the node it is attached to. Seems in the case of `TSInterfaceDeclaration` we need to read the actual location from the comment node and use that to find out where in the `TSInterfaceDeclaration` the comment was.

**Comment by vjeux** [comment]
merged

**Pull request merged**